### PR TITLE
Make possible to use custom domain name validation

### PR DESCRIPTION
**Pull request by talyssonoc**
Refs https://github.com/ambethia/recaptcha/issues/167

### CODE REVIEW

File: lib/recaptcha/verify.rb [hunk 1]
```diff
@@ -13,6 +13,7 @@ def verify_recaptcha(options = {})
 
       private_key = options[:private_key] || Recaptcha.configuration.private_key!
       recaptcha_response = options[:response] || params['g-recaptcha-response'].to_s
+      custom_domain_validation = options[:domain] || nil
```
Comment (grosser) [inline lib/recaptcha/verify.rb]: don't need to `|| nil` ... it will be nil anyway :)

Comment (talyssonoc) [inline lib/recaptcha/verify.rb]: Oh, indeed! Gonna change.

File: README.md [hunk 2]
```diff
@@ -89,13 +89,13 @@ Some of the options available:
...
-| :message     | Custom error message
+| :model       | Model to set errors.
+| :attribute   | Model attribute to receive errors. (default :base)
+| :message     | Custom error message.
 | :private_key | Override private API key.
 | :timeout     | The number of seconds to wait for reCAPTCHA servers before give up. (default `3`)
-| :response    | Custom response parameter (default: params['g-recaptcha-response'])
-
+| :response    | Custom response parameter. (default: params['g-recaptcha-response'])
+| :domain      | Domain name to check equality when you need custom [domain name validation](https://developers.google.com/recaptcha/docs/domain_validation). You can also pass a callable that receives the [hostname field](https://developers.google.com/recaptcha/docs/verify#api-response) and returns if domain name is valid or not. (default: `nil`)
```
Comment (grosser) [inline README.md]: How about:

Expected hostname or a callable that validates the hostname, see [domain validation](https://developers.google.com/recaptcha/docs/domain_validation) and [hostname](https://developers.google.com/recaptcha/docs/verify#api-response) docs.  (default: `nil`)

File: lib/recaptcha/verify.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@ def verify_recaptcha(options = {})
 
       private_key = options[:private_key] || Recaptcha.configuration.private_key!
       recaptcha_response = options[:response] || params['g-recaptcha-response'].to_s
+      custom_domain_validation = options[:domain]
```
Comment (grosser) [inline lib/recaptcha/verify.rb]: I'd go full `hostname` here ... it's closer to the api ...

File: lib/recaptcha/verify.rb [hunk 4]
```diff
@@ -23,10 +24,16 @@ def verify_recaptcha(options = {})
...
 
-        reply = Recaptcha.get(verify_hash, options)
-        answer = JSON.parse(reply)['success']
+        raw_reply = Recaptcha.get(verify_hash, options)
+        reply = JSON.parse(raw_reply)
+        answer = reply['success']
+        domain_validated = true
 
-        if answer.to_s == 'true'
+        if custom_domain_validation
```
Comment (grosser) [inline lib/recaptcha/verify.rb]: `if hostname_validation = options[:hostname]`

File: lib/recaptcha/verify.rb [hunk 5]
```diff
@@ -63,6 +70,16 @@ def verify_recaptcha!(options = {})
 
     private
 
+    def domain_validated?(hostname, custom_domain_validation)
+      if custom_domain_validation.respond_to?(:call)
```
Comment (grosser) [inline lib/recaptcha/verify.rb]: could make this simpler and let the .call blow up ... should be obvious ...

```Ruby
if custom_domain_validation.is_a?(String)
  custom_domain_validation == hostname
else
  custom_domain_validation.call(hostname)
end
```

File: test/verify_test.rb [hunk 6]
```diff
@@ -144,6 +144,43 @@
       refute @controller.verify_recaptcha
       assert_nil @controller.flash[:recaptcha_error]
     end
+
+    it "check for equality when string custom domain validation is passed" do
+      domain = 'fake.domain.com'
+
+      expect_http_post.to_return(body: %{{"success":true, "hostname": "#{domain}"}})
```
Comment (grosser) [inline test/verify_test.rb]: the stub is repeated 4 times ... could be done in a setup

**Comment by grosser** [comment]
I like it, just a few nitpicks :)

**Comment by talyssonoc** [comment]
@grosser addressed them all!

**Comment by grosser** [comment]
:heart:

**Comment by grosser** [comment]
1.2.0 is out!

had to fix 1 more bug that made it unusable ...
screen shot 2016-03-29 at 9 13 22 am

**Comment by talyssonoc** [comment]
Great! :smile:

**Pull request merged**